scripts/AudioManager.js

```diff
@@ -0,0 +1,112 @@
+// Handles playing hourly music, KK, and the town tune.
+
+'use strict';
+
+function AudioManager(addEventListener, isTownTune) {
+
+	var audio = document.createElement('audio');
+	var townTuneManager = new TownTuneManager();
+
+	// isHourChange is true if it's an actual hour change,
+	// false if we're activating music in the middle of an hour
+	function playHourlyMusic(hour, game, isHourChange) {
+		audio.loop = true;
+		audio.removeEventListener("ended", playKKSong);
+		var fadeOutLength = isHourChange ? 3000 : 500;
+		fadeOutAudio(fadeOutLength, function() {
+			if (isHourChange && isTownTune()) {
+				townTuneManager.playTune(function() {
+					audio.src = getSrc(game, hour);
+					playPause(true);
+				});
+			}
+			else {
+					audio.src = getSrc(game, hour);
+					playPause(true);
+			}
+		});
+	}
+
+	// isWeatherChange is true if it's an actual hour change,
+	// false if we're activating music in the middle of an hour
+	function playWeatherMusic(hour, game, weather, isHourChange) {
+		audio.loop = true;
+		audio.removeEventListener("ended", playKKSong);
+		var fadeOutLength = isHourChange ? 3000 : 500;
+		fadeOutAudio(fadeOutLength, function() {
+			if (isHourChange && isTownTune()) {
+				townTuneManager.playTune(function() {
+					audio.src = getSrc(game, hour, weather);
+					playPause(true);
+				});
+			}
+			else {
+					audio.src = getSrc(game, hour, weather);
+					playPause(true);
+			}
+		});
+	}
+	
+	function playKKMusic() {
+		audio.loop = false;
+		audio.addEventListener("ended", playKKSong);
+		fadeOutAudio(500, playKKSong);
+	}
+
+	function playKKSong() {
+		var randomSong = Math.floor((Math.random() * 36) + 1).toString();
+		audio.src = '../kk/' + randomSong + '.ogg';
+		playPause(true);
+	}
+
+	// Fade out audio and call callback when finished.
+	function fadeOutAudio(time, callback) {
+		if (audio.paused) {
+			if (callback) callback();
+		} else {
+			var oldVolume = audio.volume;
+			var step = audio.volume / (time / 100.0);
+			var fade = setInterval(function() {
+				if (audio.volume > step) {
+					audio.volume -= step;
+				} else {
+					clearInterval(fade);
+					playPause(false);
+					audio.volume = oldVolume;
+					if (callback) callback();
+				}
+			}, 100);
+		}
+	}
+	
+	function playPause(play) {
+		if(play) {
+			audio.play();
+			setIcon(!audio.paused);
+		}
+		else {
+			audio.pause();
+			clearBadge();
+			setIcon(!audio.paused);
+		}
+	}
+
+	addEventListener("hourMusic", playHourlyMusic);
+
+	addEventListener("weatherMusic", playWeatherMusic);
```

Comment:
I feel like it would be better to handle determining what weather's music we want to play in StateManager rather than AudioManager. For example, if it is snowing and we're on the live weather option, just send "new-leaf-snowing" as the game in a normal hourMusic message. This setup clutters the system with two new events, and means if we want to change the way music plays, we would need to do it in multiple places - and I don't see any advantages to determining what music should be played in AudioManager that offset this.